File: 55-jump-game/jump-game.java
class Solution {
    int[] nums;
    Boolean[]cache;
    private boolean dp(int index){
        if(index==nums.length-1)
            return true;
        else if(index>nums.length-1)
            return false;
        if(this.cache[index]!=null)
            return this.cache[index];
        for(int i=1;i<=nums[index];i++){
            if(dp(index+i))
                return true;
            else
                this.cache[index]=false;
        }
        return false;
    }
    public boolean canJump(int[] nums) {
        this.nums=nums;
        cache=new Boolean[nums.length];
        
        return dp(0);
    }
}

File: 55-jump-game/jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        length = len(nums)
        jump = nums[0]
        i = 0
        if length == 1:
            return True
        while i < length and jump > 0:
            if jump + i >= length:
                return True
            if nums[i] >= jump:
                jump = nums[i]
            else:
                jump -= 1
            i += 1

        return False
